# **Programming for Data Science - 21KHDL1**
# **Final Project**
# **Topic:** Life Expectancy (WHO)

---

### **Student Information**:

StudentID|Full Name
-|-
21127012|Tran Huy Ban
21127143|Nguyen Minh Quan

<br>

---

## **Table of contents**

[Overview](#overview)

[Data Collection](#collect)

[Data Pre-processing and Exploration](#pre-process_explore)

[Quickview of the data](#quickview)

[Questions](#quesion)

[Reflection](#reflection)

[References](#references)

---

## **Overview** <a name="overview"></a>

<center>
<h3>
    <b>
     Demography: learn about the population and economic situation as well as related issues of countries around the world
    </b>
</h3>
    <img style="padding:10px" src="https://www.questionpro.com/blog/wp-content/uploads/2018/04/demographic_segmentation_examples.jpg" width="450"/>
</center>
Demography is the statistical study of human populations. Demography examines the size, structure, and movements of populations over space and time. It uses methods from history, economics, anthropology, sociology, and other fields. Demography is useful for governments and private businesses as a means of analyzing and predicting social, cultural, and economic trends related to population.

---

### **Libraries used**

In [35]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

---

## **Data Collection** <a name="collect"></a>

- **Data about**: The data-set related to life expectancy, health factors for 193 countries has been collected from the same WHO data repository website and its corresponding economic data was collected from United Nation website. Among all categories of health-related factors only those critical factors were chosen which are more representative. This dataset has considered data from year 2000-2015 for 193 countries  

- **Source**: From Kaggle 

- **Collect data and License**: The project relies on accuracy of data. The Global Health Observatory (GHO) data repository under World Health Organization (WHO) keeps track of the health status as well as many other related factors for all countries and collected from the same WHO data repository website and its corresponding economic data was collected from United Nation website with the help of Deeksha Russell and Duan Wang.

In [36]:
df = pd.read_csv('Data/LifeExpectancyData.csv')
df.head()

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles  ...  \
0             62     0.01               71.279624         65.0     1154  ...   
1             64     0.01               73.523582         62.0      492  ...   
2             66     0.01               73.219243         64.0      430  ...   
3             69     0.01               78.184215         67.0     2787  ...   
4             71     0.01                7.097109         68.0     3013  ...   

   Polio  Total expenditure  Diphtheria  HIV/AIDS         GDP  Population  \
0    6.0               8.16        65.0       0.1  584.259210  33736494.0   
1   58.0               8.18        62.0       0.1  612.696514    327582.0   
2   62.0               8.13        64.0       0.1  631.744976  31731688.0   
3   67.0               8.52        67.0       0.1  669.959000   3696958.0   
4   68.0               7.87        68.0       0.1   63.537231   2978599.0   

   thinness 10-19 years  thinness 5-9 years  Income composition of resources  \
0                  17.2                17.3                            0.479   
1                  17.5                17.5                            0.476   
2                  17.7                17.7                            0.470   
3                  17.9                18.0                            0.463   
4                  18.2                18.2                            0.454   

   Schooling  
0       10.1  
1       10.0  
2        9.9  
3        9.8  
4        9.5  

[5 rows x 22 columns]

---

## **Data Pre-processing and Exploration** <a name="pre-process_explore"></a>

**1. Firstly, we should find out what our dataset contains.**

In [37]:
df.shape

(2938, 22)

So in this Dataset, we have:
- **2938 rows**
- **22 columns**

**2. What is the meaning of each row ?**

Each data line represents demographic information and details on population, demographics, and economic status of **each country** for **each year**.

**3. Does the data have duplicate rows ?**


In [38]:
duplicate_rows = df[df.duplicated()]
if duplicate_rows.empty:
    print('No duplicate rows found')
else:
    print('Duplicate rows found')
    print(duplicate_rows)


No duplicate rows found


**4. What is the meaning of each columns ?**

In [39]:
df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles', ' BMI ', 'under-five deaths', 'Polio', 'Total expenditure',
       'Diphtheria', 'HIV/AIDS', 'GDP', 'Population', 'thinness 10-19 years',
       'thinness 5-9 years', 'Income composition of resources', 'Schooling'],
      dtype='object')

In [40]:
df.rename(columns={'Life expectancy ':'Life expectancy', ' BMI ': 'BMI'}, inplace=True)

**Describe meaning of each column**

| Column                           | Describe                                                                                         |
|----------------------------------|--------------------------------------------------------------------------------------------------|
| Country                          | Name of each country                                                                             |
| Year                             | The year of data recording                                                                       |
| Status                           | State of development of the country in the year under consideration (Developed, Developing)      |
| Life expectancy                  | Average life expectancy of that country                                                          |
| Adult Mortality                  | Adult Mortality Rates of both sexes (probability of dying between 15 and 60 years per 1000 population)                        |
| Infant deaths                    | Number of Infant Deaths per 1000 population                                                      |
| Alcohol                          | Alcohol, recorded per capita (15+) consumption (in litres of pure alcohol)                       |
| Percentage expenditure           | Expenditure on health as a percentage of Gross Domestic Product per capita (%)                   |
| Hepatitis B                      | Hepatitis B (HepB) immunization coverage among 1-year-olds (%)                                   |
| Measles                          | Measles - number of reported cases per 1000 population                                           |
| BMI                              | Average Body Mass Index of entire population                                                     |
| Under-five deaths                | Number of under-five deaths per 1000 population                                                  |
| Polio                            | Polio (Pol3) immunization coverage among 1-year-olds (%)                                         |
| Total expenditure                | General government expenditure on health as a percentage of total government expenditure (%)                                |
| Diphtheria                       | Diphtheria tetanus toxoid and pertussis (DTP3) immunization coverage among 1-year-olds (%)       |
| HIV/AIDS                         | Deaths per 1 000 live births HIV/AIDS (0-4 years)                                                |
| GDP                              | Gross Domestic Product per capita (in USD)                                                       |
| Population                       | Population of the country                                                                        |
| Thinness 10-19 years             | Prevalence of thinness among children and adolescents for Age 10 to 19 (% )                      |
| Thinness 5-9 years               | Prevalence of thinness among children for Age 5 to 9(%)                                          |
| Income composition of resources  | Human Development Index in terms of income composition of resources (index ranging from 0 to 1)  |
| Schooling                        | Number of years of Schooling(years)                                                              |


**5. What is the current data type of each columns? Are there any columns having inappropriate data types ?**

In [41]:
df.dtypes

Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
BMI                                float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
HIV/AIDS                           float64
GDP                                float64
Population                         float64
thinness 10-19 years               float64
thinness 5-9 years                 float64
Income composition of resources    float64
Schooling                          float64
dtype: object

Overall, it can be seen that almost all columns have suitable data types.

**6. With each ``numerical column``, how are values distributed?**

**a. What is the percentage of missing values?**

In [42]:
numeric_df = df.select_dtypes(include=['number'])
missing_percentage = numeric_df.isnull().mean()

print('The percentage of missing values:')
for idx, missing in zip(missing_percentage.index, missing_percentage): 
    print(f'- {idx}: {round(missing * 100, 2)}%')

The percentage of missing values:
- Year: 0.0%
- Life expectancy: 0.34%
- Adult Mortality: 0.34%
- infant deaths: 0.0%
- Alcohol: 6.6%
- percentage expenditure: 0.0%
- Hepatitis B: 18.82%
- Measles: 0.0%
- BMI: 1.16%
- under-five deaths: 0.0%
- Polio: 0.65%
- Total expenditure: 7.69%
- Diphtheria: 0.65%
- HIV/AIDS: 0.0%
- GDP: 15.25%
- Population: 22.19%
- thinness 10-19 years: 1.16%
- thinness 5-9 years: 1.16%
- Income composition of resources: 5.68%
- Schooling: 5.55%


**b. Min? max? Are they abnormal?**

In [43]:
numeric_df.describe()

Year  Life expectancy  Adult Mortality  infant deaths  \
count  2938.000000      2928.000000      2928.000000    2938.000000   
mean   2007.518720        69.224932       164.796448      30.303948   
std       4.613841         9.523867       124.292079     117.926501   
min    2000.000000        36.300000         1.000000       0.000000   
25%    2004.000000        63.100000        74.000000       0.000000   
50%    2008.000000        72.100000       144.000000       3.000000   
75%    2012.000000        75.700000       228.000000      22.000000   
max    2015.000000        89.000000       723.000000    1800.000000   

           Alcohol  percentage expenditure  Hepatitis B        Measles  \
count  2744.000000             2938.000000  2385.000000    2938.000000   
mean      4.602861              738.251295    80.940461    2419.592240   
std       4.052413             1987.914858    25.070016   11467.272489   
min       0.010000                0.000000     1.000000       0.000000   
25%       0.877500                4.685343    77.000000       0.000000   
50%       3.755000               64.912906    92.000000      17.000000   
75%       7.702500              441.534144    97.000000     360.250000   
max      17.870000            19479.911610    99.000000  212183.000000   

               BMI  under-five deaths        Polio  Total expenditure  \
count  2904.000000        2938.000000  2919.000000         2712.00000   
mean     38.321247          42.035739    82.550188            5.93819   
std      20.044034         160.445548    23.428046            2.49832   
min       1.000000           0.000000     3.000000            0.37000   
25%      19.300000           0.000000    78.000000            4.26000   
50%      43.500000           4.000000    93.000000            5.75500   
75%      56.200000          28.000000    97.000000            7.49250   
max      87.300000        2500.000000    99.000000           17.60000   

        Diphtheria     HIV/AIDS            GDP    Population  \
count  2919.000000  2938.000000    2490.000000  2.286000e+03   
mean     82.324084     1.742103    7483.158469  1.275338e+07   
std      23.716912     5.077785   14270.169342  6.101210e+07   
min       2.000000     0.100000       1.681350  3.400000e+01   
25%      78.000000     0.100000     463.935626  1.957932e+05   
50%      93.000000     0.100000    1766.947595  1.386542e+06   
75%      97.000000     0.800000    5910.806335  7.420359e+06   
max      99.000000    50.600000  119172.741800  1.293859e+09   

       thinness 10-19 years  thinness 5-9 years  \
count           2904.000000         2904.000000   
mean               4.839704            4.870317   
std                4.420195            4.508882   
min                0.100000            0.100000   
25%                1.600000            1.500000   
50%                3.300000            3.300000   
75%                7.200000            7.200000   
max               27.700000           28.600000   

       Income composition of resources    Schooling  
count                      2771.000000  2775.000000  
mean                          0.627551    11.992793  
std                           0.210904     3.358920  
min                           0.000000     0.000000  
25%                           0.493000    10.100000  
50%                           0.677000    12.300000  
75%                           0.779000    14.300000  
max                           0.948000    20.700000

**7. With each ``categorical column``, how are values distributed?**

**a. What is the percentage of missing values?**

In [44]:
categorical_df = df.select_dtypes(include=['object'])
missing_percentage = categorical_df.isnull().mean()

print('The percentage of missing values:')
for idx, missing in zip(missing_percentage.index, missing_percentage): 
    print(f'- {idx}: {round(missing * 100, 2)}%')

The percentage of missing values:
- Country: 0.0%
- Status: 0.0%


**b. How many different values? Are they abnormal?**

In [45]:
for col in categorical_df.columns:
    print(f'Column: {col}')
    unique_values = np.unique(categorical_df[col])
    print(f'- Number of unique values: {len(unique_values)}')
    print(f'-Some unique values: {unique_values[:5]}\n')

Column: Country
- Number of unique values: 193
-Some unique values: ['Afghanistan' 'Albania' 'Algeria' 'Angola' 'Antigua and Barbuda']

Column: Status
- Number of unique values: 2
-Some unique values: ['Developed' 'Developing']



**9. Fill in the ``missing data`` currently blank in the dataset.**

- For the current demographic dataset, it can be observed that there are numerous columns with missing data. 

- Surprisingly, three columns, namely 'Hepatitis B,' 'GDP,' and 'Population,' exhibit a notably high rate of missing data. According to the data description, this is attributed to certain countries such as Vanuatu, Tonga, Togo, Cabo Verde, etc., not disclosing and not being extensively researched.

- Before diving into how to fill in missing data, we will learn more about the nature of the missing data for each corresponding column. In it, we will conduct tests with blank data for each column and corresponding country. Then count the number of data blanks to know the number of years missing data for the corresponding column and country to calculate the method.

- In this section, we only consider these 3 columns because the data in these 3 columns has a too high missing rate and we will focus on these 3 columns. For the other columns, although we do not consider them too deeply, they are I will still fill them.

**Checking "Hepatitis B" column**

In [46]:
countries_with_null_hepatitis_B = df.loc[df['Hepatitis B'].isnull(), 'Country']
countries_count_hepatitis_B = countries_with_null_hepatitis_B.value_counts()
pd.set_option('display.max_rows', None)
print(countries_count_hepatitis_B)

Country
Denmark                                                 16
United Kingdom of Great Britain and Northern Ireland    16
Norway                                                  16
Japan                                                   16
Switzerland                                             16
Finland                                                 16
Slovenia                                                16
Iceland                                                 16
Hungary                                                 16
South Sudan                                             15
Equatorial Guinea                                       14
Somalia                                                 13
Haiti                                                   13
Sweden                                                  11
Netherlands                                             11
Namibia                                                 10
Ireland                                         

**Checking "GDP" column**

In [47]:
countries_with_null_gdp = df.loc[df['GDP'].isnull(), 'Country']
countries_count_gdp = countries_with_null_gdp.value_counts()
pd.set_option('display.max_rows', None)
print(countries_count_gdp)

Country
Bahamas                                                 16
Micronesia (Federated States of)                        16
Viet Nam                                                16
Venezuela (Bolivarian Republic of)                      16
United States of America                                16
United Republic of Tanzania                             16
United Kingdom of Great Britain and Northern Ireland    16
The former Yugoslav republic of Macedonia               16
Slovakia                                                16
Saint Vincent and the Grenadines                        16
Saint Lucia                                             16
Republic of Moldova                                     16
Republic of Korea                                       16
Bolivia (Plurinational State of)                        16
Yemen                                                   16
Gambia                                                  16
Iran (Islamic Republic of)                      

**Checking "Population" column**

In [48]:
countries_with_null_population = df.loc[df['Population'].isnull(), 'Country']
countries_count_population = countries_with_null_population.value_counts()
pd.set_option('display.max_rows', None)
print(countries_count_population)

Country
Antigua and Barbuda                                     16
Singapore                                               16
New Zealand                                             16
Oman                                                    16
Qatar                                                   16
Republic of Korea                                       16
Republic of Moldova                                     16
Saint Lucia                                             16
Saint Vincent and the Grenadines                        16
Saudi Arabia                                            16
Slovakia                                                16
Libya                                                   16
Somalia                                                 16
The former Yugoslav republic of Macedonia               16
United Arab Emirates                                    16
United Kingdom of Great Britain and Northern Ireland    16
United Republic of Tanzania                     

- In short, it can be seen that, in case of missing data, there will be many different scenarios. As this involves data from each country from 2000 to 2015, over these 16 years some countries are only missing a few data points, while others may be missing entire columns throughout. 16 years.

- To fill in these missing values, we suggest using the following methods:

- First: for countries that have columns of missing values ​​for only a few years and have values ​​in the corresponding attribute columns in later/previous years, we will use backward and forward fill for the cells This is to ensure that the error rate is not too significant.

- Second: use the method to fill the remaining blank cells with the median value. Here, these blank cells will be the cells for countries with no information during the 16 years of testing.

- The reason for this is that instead of using the median method from the beginning, it will not completely ensure the accuracy of the data and can lead to large errors. Therefore, using backward and forward will be more or less accurate in some cells.

---

**Fill NaN values**

In [49]:
country_list = df.Country.unique() # get country in dataset

# get list columns to fill missing value
fill_list = ['Life expectancy','Adult Mortality','Alcohol','Hepatitis B','BMI','Polio',
            'Total expenditure','Diphtheria','GDP','Population','thinness 10-19 years',
            'thinness 5-9 years','Income composition of resources','Schooling']

**Filling using Forward and Backward**

In [50]:
for country in country_list:
    for col in fill_list:
        
        #fill missing value with backward and forward
        df.loc[df['Country'] == country, col] = df.loc[df['Country'] == country, col].fillna(method='ffill').fillna(method='bfill')

- After this step, countries with fewer than 16 data points in a property column will be filled (if sufficient data is available for 16 years). Additionally, countries with only 1 or a few data points cannot be filled.

- Then, we will fill another missing value cell by using Median value.

**Filling using Median value**

In [54]:
for col in fill_list:
    df[col].fillna(df[col].median(), inplace=True)

In [55]:
# Verifying null-values after applying above methods.
df.isnull().sum()

Country                            0
Year                               0
Status                             0
Life expectancy                    0
Adult Mortality                    0
infant deaths                      0
Alcohol                            0
percentage expenditure             0
Hepatitis B                        0
Measles                            0
BMI                                0
under-five deaths                  0
Polio                              0
Total expenditure                  0
Diphtheria                         0
HIV/AIDS                           0
GDP                                0
Population                         0
thinness 10-19 years               0
thinness 5-9 years                 0
Income composition of resources    0
Schooling                          0
dtype: int64

## **Quickview of the data** <a name="quickview"></a>

// Draw some figures...

---

## **Questions** <a name="question"></a>

#### After having an overview of the data, we see a number of issues that need attention:
- Question 1:
- Question 2:
- Question 3:
- Question 4:
- Question 5:
- Question 6:

**Question 1: ...**
- **Meaning**
- **Idea**

In [84]:
##TODO

**Question 2: ...**
- **Meaning**
- **Idea**

In [85]:
##TODO

**Question 3: ...**
- **Meaning**
- **Idea**

In [86]:
##TODO

**Question 4: ...**
- **Meaning**
- **Idea**

In [87]:
##TODO

**Question 5: ...**
- **Meaning**
- **Idea**

In [88]:
##TODO

**Question 6: ...**
- **Meaning**
- **Idea**

In [89]:
##TODO

---

## **Reflection** <a name="reflection"></a>

#### **Difficulties encountered:**
- Tran Huy Ban:
    - a
    - b
    - c
- Nguyen Minh Quan:
    - a
    - b
    - c

#### **Knowledge learned:**
- Tran Huy Ban:
    - a
    - b
    - c
- Nguyen Minh Quan:
    - a
    - b
    - c

#### **What would we do if we had more time?**
-
-
-

---

## **References** <a name="references"></a>